ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -288,15 +314,20 @@ localRoots actions@PeerSelectionActions{ readLocalRootPeers
                                        }
            policy
            st@PeerSelectionState{
-             bigLedgerPeers,
              localRootPeers,
              publicRootPeers,
              knownPeers,
              establishedPeers,
              activePeers,
              inProgressDemoteHot,
-             targets = PeerSelectionTargets{targetNumberOfKnownPeers}
-           } =
+             targets = PeerSelectionTargets{targetNumberOfKnownPeers},
+             currentBootstrapPeersFlag,
+             currentLedgerStateJudgement
+           }
+  -- If the current ledger state is TooOld we can only trust our local
+  -- root peers, this means that if we remove any local root peer we
+  -- no longer
```

Comment:
> This code is also not right. If we are using bootstrap peers, we should only retain trusted local roots.

We do the simpler thing to preserve the invariant and disconnect from everybody, before connecting to only trusted peers. I will ask them, but I think finer granularity on local roots is desirable, we will always have to clamp to the limit, otherwise we risk breaking an invariant

Reply:
I consulted Neil and Karl they said that `trusted` flag should be at the same level of `valency` and `advertise` which I agree. The code however distributes these values throughout all peers in a group

Reply:
re: "This code is also not right. If we are using bootstrap peers, we should only retain trusted local roots."

You are right, but I will fix that in a different commit, since in this trust-ability flag hasn't been introduced.